# Testing the sklearn Elastic Nets implementation

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import numpy as np
import pandas as pd

In [3]:
import scanpy.api as sc

sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)
sc.settings.set_figure_params(dpi=80, color_map='viridis')  # low dpi (dots per inch) yields small inline figures
sc.logging.print_versions()

scanpy==1.3.2+25.g8ac9c03 anndata==0.6.11 numpy==1.14.6 scipy==1.1.0 pandas==0.23.4 scikit-learn==0.20.0 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 


In [4]:
import sys
sys.path.append('/home/ahsvargo/xvalid')
sys.path.append('/home/ahsvargo/.local/bin')

In [5]:
from picturedrocks import Rocks
from picturedrocks.performance import FoldTester, PerformanceReport, NearestCentroidClassifier

In [6]:
import sklearn.linear_model as sklm

In [7]:
import scipy.sparse as spsp

## Load the data

In [8]:
adata = sc.datasets.paul15()

... storing 'paul15_clusters' as categorical


In [9]:
lookup = list(adata.obs['paul15_clusters'].cat.categories)
yVec = np.array([lookup.index( adata.obs['paul15_clusters'][i] ) for i in range(adata.obs['paul15_clusters'].shape[0]) ])

## Make an elastic net

This one did not finish running after about 5 hours...  look for timing of one elastic net

In [10]:
vec0 = (yVec == 0)*1

In [19]:
en = sklm.ElasticNetCV(l1_ratio=[.1, .5, .7, .9, .95, .99, 1], cv=5, max_iter=10000)

In [ ]:
%%time
en.fit(adata.X, vec0)

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did n

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.



In [11]:
en = sklm.ElasticNet(max_iter=10000)

In [13]:
%%time
en.fit(adata.X, vec0)

CPU times: user 106 ms, sys: 105 ms, total: 212 ms
Wall time: 1.17 s


ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=10000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)

Hmm... that seems fast enough that we could have run the CV manually.  I would predict only 70 minutes based on this time (with 7 values of the l1_ratio and 100 values of alpha to test and 5 folds).

In [14]:
(1.2 * 7 * 100 * 5)/60

70.0

In [17]:
np.nonzero(en.coef_)

(array([553]),)

In [19]:
%%time
en.predict(adata.X)

CPU times: user 34.1 ms, sys: 255 ms, total: 289 ms
Wall time: 108 ms


array([0.01387042, 0.01372338, 0.02048741, ..., 0.01416451, 0.01387042,
       0.02489875], dtype=float32)

Just one with alpha=1.  Wonder how that corresponds to the markers that we select... Check if time

In [45]:
%%time
en = sklm.ElasticNet(l1_ratio=0.1,max_iter=10000, alpha=0.04, tol=0.0001)
en.fit(adata.X, vec0)

CPU times: user 22.9 s, sys: 119 ms, total: 23 s
Wall time: 22.8 s


In [26]:
0.05*1000

50.0

In [17]:
0.1 * 0.02

0.002

In [18]:
0.9*0.02

0.018000000000000002

In [28]:
adata.X.shape

(2730, 3451)

In [37]:
np.nonzero(en.coef_)[0].shape

(1,)

Smaller values of alpha tend to take longer it appears (and result in more features selected).  So perhaps the CV method is exploring a list of smaller and smaller values of alpha.  At this rate (taking all 10000 iterations) we would take much longer to complete the CV - about 40 hours, it appears:

In [38]:
(41*100*7*5)/60/60

39.86111111111111

What about reducing the number of iterations?  We didn't converge anyway...

Seems to be approximately linear in the number of iterations when we don't converge (which makes some sense).  So... that would be four hours per cluster for each fold.  Pretty much the same as D3E, except that we are using fewer cores.  So I could run it, at least on these smaller data sets.

In [47]:
%%time
en = sklm.ElasticNet(l1_ratio=0.1,max_iter=1000, alpha=0.01)
en.fit(adata.X, vec0)

CPU times: user 3.83 s, sys: 126 ms, total: 3.96 s
Wall time: 3.76 s


/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.



In [49]:
np.nonzero(en.coef_)[0].shape

(744,)

A question here: what do we do with the fact that this only returns a list of markers?  Or should I be looking at different values of the `l1_ratio` as a proxy for the sparsity parameter in our 1bcs model? 

Specifically, what do we do if we only get a small number of markers for one of the clusters?

It appears that we dynamically create the list of alphas when we try to fit the model, however.  I can't really control this.  I can set the list of alphas if I desire, but that seems to take away a bit from the point of this method.  We can automatically use multiple cores though.  That's really nice.

In [54]:
en = sklm.ElasticNetCV(l1_ratio=[.1, .5, .7, .9, .95, .99, 1], cv=5, max_iter=1000, n_jobs=2)

In [25]:
en.alphas_

AttributeError: 'ElasticNet' object has no attribute 'alphas_'

What if we try sparse data?  The method seems to take significantly longer ~ 3 minutes in this test, without converging.  I tried as both `spsp.csc_matrix` and `spsp.csr_matrix`.  

So we can come up with our own array of alphas.  That's fine

In [39]:
spX = spsp.coo_matrix(adata.X)

In [40]:
spX.eliminate_zeros()

In [44]:
spX = spX.tocsr()

In [46]:
%%time
en = sklm.ElasticNet(l1_ratio=0.1,max_iter=10000, alpha=0.01)
en.fit(spX, vec0)

CPU times: user 2min 56s, sys: 44.2 ms, total: 2min 56s
Wall time: 2min 56s


/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.



In [45]:
type(spX)

scipy.sparse.csr.csr_matrix

From testing: I get convergence with `l1_ratio = 0.1` and `alpha = 0.05`.  This results in 108 markers, which should be enough to consider for the analysis.  We will just rock a linear list of alphas using the specified range.  We don't look at values of alpha as large as the method in the CV class normally would, as  a linear list would puts too much interest in the super large values of alpha (which seem to return a very low number of markers - 0 markers for alpha > 10).

In [38]:
alphaList = np.linspace(0.05, 100*0.05, num=100, endpoint=True)

In [41]:
en = sklm.ElasticNetCV(l1_ratio=[.1, .5, .7, .9, .95, .99, 1], cv=5, max_iter=10000, n_jobs=3, alphas=alphaList)

In [42]:
%%time
en.fit(adata.X, vec0)

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.

/home/ahsvargo/miniconda3/envs/r35py37/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.



CPU times: user 3min 32s, sys: 2min 5s, total: 5min 37s
Wall time: 2min 8s


ElasticNetCV(alphas=array([0.05, 0.1 , ..., 4.95, 5.  ]), copy_X=True, cv=5,
       eps=0.001, fit_intercept=True,
       l1_ratio=[0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1], max_iter=10000,
       n_alphas=100, n_jobs=3, normalize=False, positive=False,
       precompute='auto', random_state=None, selection='cyclic',
       tol=0.0001, verbose=0)

In [44]:
np.nonzero(en.coef_)[0]

array([ 552,  553, 1212, 1351, 1421, 2025, 2384])